Expand token security considerations (#45)
tfpauly/privacy-proxy: draft-private-access-tokens.md

Private Access Tokens and for Origins to work. However, similar DoS vectors already exist
for Origins, e.g., at the underlying TLS layer.

> ## Man-in-the-Middle Attacks

Reviewer: ```suggestion
## Channel Security
```

> An attacker that can act as a man-in-the-middle for Mediator/Issuer communication can

Reviewer: ```suggestion
An attacker that can act as a intermediate Mediator or Issuer communication via impersonation or any other means can
```

Author: ```suggestion
An attacker that can act as an intermediate between Mediator and Issuer communication can
```

Reviewer: 🤦

influence or disrupt the ability for the Issuer to correctly rate-limit token issuance.

> Communication between Mediators and Issuers MUST use a secure HTTPS connection, and SHOULD
> prevent man-in-the-middle attacks by employing an approach like TLS certificate pinning.

Reviewer: ```suggestion
All communication channels MUST use server-authenticated HTTPS. Where appropriate, e.g., between
Clients and Mediators, connections MAY mutually authenticate both client and server, or use mechanisms
such as TLS certificate pinning, to mitigate the risk of channel compromise.
```

> An attacker than can act as a man-in-the-middle for Client/Origin communication can

Reviewer: ```suggestion
An attacker than can intermediate the channel between Client and Origin can
```

observe a TokenChallenge, and can view a Token being presented for authentication

> to an Origin. Origins can avoid this Token being used for some future connection
> by the attacker impersonating the Client by ensuring that the redemption_nonce
> presented in the TokenChallenge is bound to the specific TLS session with the
> Client.

Reviewer: ```suggestion
to an Origin. The protocol in this document does not allow Origins to protect against this type of attack.

[OPEN ISSUE: unless both client and server independently derive the challenge from the underlying channel, this problem will always remain.]
```

Author: @chris-wood hm, I'd like to have a bit of a better answer here rather than declaring we can't protect against it.

The main point, from the issue, was not to prevent the client-origin communication from having an intermediary watching things, but to stop the intermediary from harvesting tokens that it could then use in _other_ connections/sessions/contexts as its own client.

If the origin can require that the challenge/nonce is specific to a given session, can't it prove that this wasn't a token harvested on behalf of an unrelated connection?

Reviewer: > If the origin can require that the challenge/nonce is specific to a given session, can't it prove that this wasn't a token harvested on behalf of an unrelated connection?

When the connection is MITM'd, the session is with the attacker, not the client, so this just doesn't seem to work. A workaround would be to derive the nonce _from_ the connection, a la via [the Export API](https://datatracker.ietf.org/doc/html/rfc8446#section-7.5), but that's pretty uncommon. Things like WebAuthn have [slots for this type of binding](https://www.w3.org/TR/webauthn-2/#dom-collectedclientdata-tokenbinding) but it's basically never used in practice.

Author: Right, the session is with the attacker, and then the client behind the attacker. Ah, so you’re saying the attacker could have other clients that it’s trying to squash together into one session?

Reviewer 2: Do I understand correctly that this is a machine-on-the-side attacker, not a MITM attacker? AIUI, the attacker:
- watches the challenge and the token go by, harvesting both; and
- later spoofs a client to reach the origin and uses a token that matches the origin's challenge.

If my understanding is correct, then the crux of the issue is the risk that the nonce might repeat (which makes it a bad nonce.)

If so, it seems to me that @tfpauly's text suggests that an Origin can make a strong nonce by deriving it from something unique to the session as seen by the Origin (for example: timestamp + H(session key) + H(some client info)). I could of course be misunderstanding the issue and the text :-)

Reviewer: @tfpauly I updated my suggestion to try and make the limitation (and open problem) more clear.

Reviewer: @janaiyengar the attacker needs to intercept TLS to see the challenge and resulting Token, so it's not enough to be on the side, but maybe I misunderstood you?

Reviewer 3: Does the extra ability to present multiple tokens in the same session, versus the attacker just mitming the individual connections and intercepting/replaying the tokens separately there introduce much of an issue? Though you can sort of solve both with the exporter secret solution from the TLS session.

# Privacy Considerations {#privacy-considerations}

## Origin Verification

Private Access Tokens are defined in terms of a Client authenticating to an Origin, where
the "origin" is used as defined in {{?RFC6454}}. In order to limit cross-origin correlation,
Clients MUST verify that the origin_name presented in the TokenChallenge structure ({{scheme}})
matches the origin that is providing the HTTP authentication challenge, where the matching logic
is defined for same-origin policies in {{?RFC6454}}. Clients MAY further limit which
authentication challenges they are willing to respond to, for example by only accepting
challenges when the origin is a web site to which the user navigated.

## Client Identification with Unique Keys

Client activity could be linked if an Origin and Issuer collude to have unique keys targeted

> at specific Clients or sets of Clients. In order to mitigate this risk, the Mediator is able
> to see and validate the key_id presented by the Client to the Issuer. As described in

Reviewer: ```suggestion
at specific Clients or sets of Clients. In order to mitigate this risk, the Mediator can 
observe and validate the key_id presented by the Client to the Issuer. As described in
```

> {{issuance}}, Mediators MUST validate that the key_id in the client's AccessTokenRequest

Reviewer: ```suggestion
{{issuance}}, Mediators MUST validate that the key_id in the Client's AccessTokenRequest
```

> matches a known public key for the Issuer. This validation should allow for key rotation across

Reviewer: ```suggestion
matches a known public key for the Issuer. This validation allows for key rotation across
```

Private Access Tokens and for Origins to work. However, similar DoS vectors already exist
for Origins, e.g., at the underlying TLS layer.

## Channel Security

> An attacker that can act as an intermediate between Mediator and Issuer communication can

Reviewer: Intermediation is what happens when the attacker is in the middle and is the thing we're concerned with here, so I'd be inclined to stick with what's here.

influence or disrupt the ability for the Issuer to correctly rate-limit token issuance.
All communication channels MUST use server-authenticated HTTPS. Where appropriate, e.g., between
Clients and Mediators, connections MAY mutually authenticate both client and server, or use mechanisms

> such as TLS certificate pinning, to mitigate the risk of channel compromise.

Reviewer: Network attackers can also just block connections to mediators or issuers, so I'm not sure client auth helps for that particular problem.

> such as TLS certificate pinning, to mitigate the risk of channel compromise.

Reviewer 3: The concern might be if a party pretending to be the mediator asserts something about some client ID causing the issuer to think they've hit some policy limit. Though similarly a malicious mediator can do the same.

> such as TLS certificate pinning, to mitigate the risk of channel compromise.

Author: Yeah, I don't know if we want specific client auth, beyond however the client would naturally authenticate to the mediator (device attestation, etc).

presented in the TokenChallenge is bound to the specific TLS session with the
Client.

# Privacy Considerations {#privacy-considerations}

## Origin Verification

Private Access Tokens are defined in terms of a Client authenticating to an Origin, where
the "origin" is used as defined in {{?RFC6454}}. In order to limit cross-origin correlation,
Clients MUST verify that the origin_name presented in the TokenChallenge structure ({{scheme}})
matches the origin that is providing the HTTP authentication challenge, where the matching logic
is defined for same-origin policies in {{?RFC6454}}. Clients MAY further limit which
authentication challenges they are willing to respond to, for example by only accepting
challenges when the origin is a web site to which the user navigated.

## Client Identification with Unique Keys

Client activity could be linked if an Origin and Issuer collude to have unique keys targeted
at specific Clients or sets of Clients. In order to mitigate this risk, the Mediator can 
observe and validate the key_id presented by the Client to the Issuer. As described in
{{issuance}}, Mediators MUST validate that the key_id in the Client's AccessTokenRequest

> matches a known public key for the Issuer. This validation allows for key rotation across

Reviewer 3: Probably as a separate issue this probably needs to be more nuanced. The issuer could decide to do a key rotation when a specific user/event happens (tracking the first 100 users vs the next set of user).

Mediators may almost need their own key rotation policy (or they delegate that to whatever known key service they use).

Author: Modified the text a bit on Jana's suggestion, but yes, we can continue to improve it.

> all Clients, but not allow for per-Client targetted keys.

Author: Moved the key rotation down to privacy considerations
